File: 2015-Average Height of Buildings in Each Segment.py
class Solution(object):
    def averageHeightOfBuildings(self, buildings):
        """
        :type buildings: List[List[int]]
        :rtype: List[List[int]]
        """
        points = []
        for x, y, h in buildings:
            points.append([x, 1, h])
            points.append([y, -1, h])
            
        points.sort()
        
        res = []
        count = heightSum = prevX = 0
        
        for i in range(len(points)):
            currX, kind, height = points[i]
            if count != 0 and prevX != currX:
                h = heightSum//count
                if res and res[-1][1] == prevX and res[-1][2] == h:
                    res[-1][1] = currX
                else:
                    res.append([prevX, currX, h])
            prevX = currX
            heightSum += kind*height
            count += kind
            
        return res

File: JavaScript/2015-Average Height of Buildings in Each Segment.js
/**
 * @param {number[][]} buildings
 * @return {number[][]}
 */
var averageHeightOfBuildings = function(buildings) {
    let points = new Array();
    
    for (let [start, end, height] of buildings){
        points.push([start, 1, height]);
        points.push([end, -1, height]);
    }
    
    points.sort(function(a,b){return a[0]-b[0];})
    let res = [];
    
    let currH = 0, count = 0, prevX = 0;
    for (let [currX, kind, height] of points){
        if (count!==0 && prevX!==currX){
            let avgH = Math.floor(currH/count);
            if (res.length && res[res.length-1][1] === prevX && avgH === res[res.length-1][2]){
                res[res.length-1][1] = currX;
            }else{
                res.push([prevX, currX, avgH]);
            }
        }
        count += kind;
        currH += kind*height;
        prevX = currX;
    }
    
    return res;
};